### PR TITLE
[receiver/skywalking] add Skywalking tracing receiver impl

### PR DESCRIPTION
**Pull request by JaredTan95**
**Description:** add Skywalking tracing receiver impl

1. impl skywalking grpc/http channel to receiver skywalking agent tracing data.
2. dummy response skywalking's grpc service.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7954

**Testing:** tested with following steps:
1.  otel config yaml:
```yaml
receivers:
  # The following demonstrates how to enable protocols with defaults.
  skywalking/defaults:
    protocols:
      grpc:
  skywalking:
    protocols:
      grpc:
        endpoint: 0.0.0.0:11800
      http:
        endpoint: 0.0.0.0:12800
  skywalking/customname:
    protocols:
      http:
        endpoint: 0.0.0.0:12801

exporters:
  jaeger:
    endpoint: "jaeger:14250"
    tls:
      insecure: true

processors:
  batch:

service:
  pipelines:
    traces:
      receivers: [skywalking]
      processors: [batch]
      exporters: [jaeger]

```
3. jaeger ui tracing view:
image
image

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by dashpole** [comment]
cc @jpkrohling, who I believe is sponsoring the component

File: receiver/skywalkingreceiver/skywalkingproto_to_traces.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+var OtSpanTagsMapping = map[string]string{
+	"url":         conventions.AttributeHTTPURL,
+	"status_code": conventions.AttributeHTTPStatusCode,
+	"db.type":     conventions.AttributeDBSystem,
+	"db.instance": conventions.AttributeDBName,
+	"mq.broker":   conventions.AttributeNetPeerName,
+}
+
+func SkywalkingToOtlpTraces(segment *agentV3.SegmentObject) pdata.Traces {
```
Comment (jpkrohling) [inline receiver/skywalkingreceiver/skywalkingproto_to_traces.go:38]:
```suggestion
func SkywalkingToTraces(segment *agentV3.SegmentObject) pdata.Traces {
```

or 

```suggestion
func SkywalkingToOTLPTraces(segment *agentV3.SegmentObject) pdata.Traces {
```

File: receiver/skywalkingreceiver/skywalkingproto_to_traces.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+	"time"
+	"unsafe"
+
+	"go.opentelemetry.io/collector/model/pdata"
+	conventions "go.opentelemetry.io/collector/model/semconv/v1.8.0"
+	common "skywalking.apache.org/repo/goapi/collect/common/v3"
+	agentV3 "skywalking.apache.org/repo/goapi/collect/language/agent/v3"
+)
+
+var OtSpanTagsMapping = map[string]string{
```
Comment (jpkrohling) [inline receiver/skywalkingreceiver/skywalkingproto_to_traces.go:30]: Does it need to be public?

File: receiver/skywalkingreceiver/skywalkingproto_to_traces.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+	for _, tag := range tags {
+		otKey, ok := OtSpanTagsMapping[tag.Key]
+		if ok {
+			attrs.UpsertString(otKey, tag.Value)
+		}
+	}
+}
+
+func swSpansToOtlpSpans(traceId string, spans []*agentV3.SpanObject, dest pdata.SpanSlice) {
```
Comment (jpkrohling) [inline receiver/skywalkingreceiver/skywalkingproto_to_traces.go:81]:
```suggestion
func swSpansToSpanSlice(traceId string, spans []*agentV3.SpanObject, dest pdata.SpanSlice) {
```

or 

```suggestion
func swSpansToOTLPSpans(traceId string, spans []*agentV3.SpanObject, dest pdata.SpanSlice) {
```

File: receiver/skywalkingreceiver/skywalkingproto_to_traces.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+	dest.EnsureCapacity(len(spans))
+	for _, span := range spans {
+		if span == nil {
+			continue
+		}
+		swSpanToOtelSpan(traceId, span, dest.AppendEmpty())
+	}
+}
+
+func swSpanToOtelSpan(traceId string, span *agentV3.SpanObject, dest pdata.Span) {
```
Comment (jpkrohling) [inline receiver/skywalkingreceiver/skywalkingproto_to_traces.go:95]:
```suggestion
func swSpanToSpan(traceId string, span *agentV3.SpanObject, dest pdata.Span) {
```

or

```suggestion
func swSpanToOTLPSpan(traceId string, span *agentV3.SpanObject, dest pdata.Span) {
```

File: receiver/skywalkingreceiver/skywalkingproto_to_traces.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+	case span.GetSpanType() == agentV3.SpanType_Entry:
+		dest.SetKind(pdata.SpanKindServer)
+	case span.GetSpanType() == agentV3.SpanType_Local:
+		dest.SetKind(pdata.SpanKindInternal)
+	default:
+		dest.SetKind(pdata.SpanKindUnspecified)
+	}
+
+	swLogsToSpanEvents(span.GetLogs(), dest.Events())
+	// swkyalking: In the across thread and across process, these references targeting the parent segments.
```
Comment (jpkrohling) [inline receiver/skywalkingreceiver/skywalkingproto_to_traces.go:136]:
```suggestion
	// skywalking: In the across thread and across processes, these references target the parent segments.
```

File: receiver/skywalkingreceiver/skywalkingproto_to_traces_test.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+package skywalkingreceiver
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSwProtoToTraces(t *testing.T) {
```
Comment (jpkrohling) [inline receiver/skywalkingreceiver/skywalkingproto_to_traces_test.go:23]: There are way too many other things there that deserve to be tested.

File: receiver/skywalkingreceiver/sw_dummy_clr_service.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+package skywalkingreceiver
+
+import (
+	"context"
+
+	common "skywalking.apache.org/repo/goapi/collect/common/v3"
+	agent "skywalking.apache.org/repo/goapi/collect/language/agent/v3"
+)
+
+type clrService struct {
```
Comment (jpkrohling) [inline receiver/skywalkingreceiver/sw_dummy_clr_service.go:24]: Are those dummy services going to be implemented?

Comment (JaredTan95) [reply receiver/skywalkingreceiver/sw_dummy_clr_service.go:24]: no, we don't need to impl those dummy services.

File: receiver/skywalkingreceiver/skywalkingproto_to_traces_test.go (hunk 6 above)
Comment (JaredTan95) [reply receiver/skywalkingreceiver/skywalkingproto_to_traces_test.go:23]: okay, I will add more test.

File: receiver/skywalkingreceiver/skywalkingproto_to_traces.go (hunk 2 above)
Comment (JaredTan95) [reply receiver/skywalkingreceiver/skywalkingproto_to_traces.go:30]: no, I will update it.

**Review by jpkrohling — APPROVED** [review]

**Comment by caoheyang** [comment]
Hi Jared, 

I have a quick question, when I want to use skywalking tracing receiver, should I use opentelemetry agent for java  or skywalking agent for java?

Thanks,

Heyang

**Comment by JaredTan95** [comment]
> 

skywalking agent

**Comment by caoheyang** [comment]
> 

Thanks, Jared,
So ,how to use opentelemetry-collector receiver the skywalking trace info? 
export anothor port for   opentelemetry-collecto?

Heyang

**Comment by JaredTan95** [comment]
As described by https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/8549#issue-1173568656:

```yaml
receivers:
  skywalking:
    protocols:
      grpc:
        endpoint: 0.0.0.0:11800
      http:
        endpoint: 0.0.0.0:12800

exporters:
  jaeger:
    endpoint: "jaeger:14250"
    tls:
      insecure: true

processors:
  batch:

service:
  pipelines:
    traces:
      receivers: [skywalking]
      processors: [batch]
      exporters: [jaeger]
```

You shuold add `skywalking` receiver, and then, you can change the `Skywalking OAP Address` to `otel collecotr 11800` port in skywalking agent config.

**Comment by caoheyang** [comment]
OKay, Thanks a lot ,Jared, I will try it.

Regards,
Heyang

**Pull request merged**